### PR TITLE
Drop type comments entirely

### PR DESCRIPTION
**Pull request by greg0ire**
This feature is only useful for reverse-engineering purposes since we
have platform-aware comparison. Reverse-engineering is not enough to
justify having that feature, and ORM metadata reverse-engineered from
the database can still be adjusted afterwards.

### CODE REVIEW

**Comment by greg0ire** [comment]
:thinking: I'm a bit surprised by the test suite result @morozov … it says there is a diff for a guid column on MariaDB, but after https://github.com/doctrine/dbal/pull/4746 , that should not be the case anymore, right?

**Comment by greg0ire** [comment]
Debugged it, and it's just because of the comment of course :)
I think the one who has it is the `$offlineTable`

**Comment by greg0ire** [comment]
The remaining failures are with IBM DB2 and SQL Server, I don't know how to debug them, so I'm going to create the ORM and migration PRs instead

**Comment by morozov** [comment]
> 🤔 I'm a bit surprised by the test suite result … it says there is a diff for a guid column on MariaDB, but after #4746, that should not be the case anymore, right?

Which test are you referring to?

**Comment by morozov** [comment]
The test on SQL Server fails because of this:
https://github.com/doctrine/dbal/blob/d78267cefb5f3145e43e5bc24f7f0d93f1ca7998/src/Platforms/SQLServerPlatform.php#L510-L513

You replaced the call to `$this->getColumnComment($column)` with `$column->getComment()` in `AbstractPlatform` but not here. So when processing the diff, the platform thinks that the deployed column has a comment, tries to update it and fails because it doesn't exist:
```sql
EXEC sp_updateextendedproperty N'MS_Description', N'foo', N'SCHEMA', 'dbo', N'TABLE', 'sqlsrv_column_comment', N'COLUMN', commented_type
-- SQLSTATE [42000, 15217]: [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Property cannot be updated or deleted. Property 'MS_Description' does not exist for 'dbo.sqlsrv_column_comment.commented_type'.
```

The tests on IBM DB2 fail because ... surprise ... it doesn't support boolean columns, as well as ~~most of the other platforms~~ MySQL and Oracle. But the same test doesn't fail say on MySQL because of this hack: https://github.com/doctrine/dbal/blob/dfdb5fb2b2230237bf46761140ce9b2dd4edea39/src/Platforms/AbstractMySQLPlatform.php#L1052

After some more research, of the two `ComparatorTest::testDefaultValueComparison()` failures on IBM DB2, only one is expected because even if the type is introspected incorrectly, the resulting SQL should be the same but it's slightly different:
```diff
- SMALLINT DEFAULT '0' NOT NULL
+ SMALLINT DEFAULT 0 NOT NULL
```
I would consider it a separate issue (https://github.com/doctrine/dbal/pull/5108). ~~Likely, similar to MariaDB, IBM DB2 provides an SQL expression representing the default values (the string `0` representing a numeric literal) while the DBAL considers it to be _the_ value (i.e. a string).~~ This is where the quotes are added: https://github.com/doctrine/dbal/blob/a4da59dea2f96076173a42ba9d16c96b701fd6f3/src/Platforms/AbstractPlatform.php#L2599-L2601

**Comment by greg0ire** [comment]
> Which test are you referring to?

Sorry, I was referring to

https://github.com/doctrine/dbal/blob/4d66064c8bf5a2a1567c74f8565f7c3b9ccc5a64/tests/Functional/Schema/MySQLSchemaManagerTest.php#L358-L371

**Comment by mvorisek** [comment]
No! Please do not remove DC2Type comments!

Following post https://github.com/doctrine/dbal/issues/5194#issuecomment-1018799512 .

a) Different DBAL Types, even if stored in the same DB column type, can have the values encoded differently. Without DC2Type comments, it is impossible to distinguish between such data/DB state and diff/migrate them correctly.

b) We use the DC2Type comments also in DB triggers to further enforce the data integrity. Without DC2Type comments, it is impossible to distinguish between different DBAL Types on the server side.

In summary, the DC2Type comments provide a very valuable and irreplaceable metadata stored together with the column definition, and they should stay forever.

**Comment by morozov** [comment]
The DBAL maintains column comments for its internal purposes which is irrelevant as of https://github.com/doctrine/dbal/pull/4746.

If the application needs to maintain column comments for its own purposes, it should do so by its own means.

**Comment by greg0ire** [comment]
@morozov I'm not sure how to interpret the remaining failures… `diffTable` returns a non-empty diff, and that makes sense since the detected type is no longer the same as the expected one. Later on, when using `toSql()` on the schema diff that would know about this table diff, the difference should vanish, right? Should `Doctrine\DBAL\Tests\Functional\Schema\ComparatorTest` just be removed? Or should it maybe be rewritten to ensure that indeed, `SchemaDiff::toSql()` is empty?

**Comment by morozov** [comment]
> `diffTable` returns a non-empty diff, and that makes sense since the detected type is no longer the same as the expected one.

It does and it doesn't. This is a remainder of the issue described in https://github.com/doctrine/dbal/pull/5107#issuecomment-991715998 and partially fixed in https://github.com/doctrine/dbal/pull/5108.

DB2 doesn't support boolean columns natively, so the DBAL emulates them via SMALLINT. The same applies to MySQL and Oracle. During schema introspection, the DBAL converts non-boolean columns back to boolean to satisfy the comparator:
1. https://github.com/doctrine/dbal/blob/dfdb5fb2b2230237bf46761140ce9b2dd4edea39/src/Platforms/AbstractMySQLPlatform.php#L1052
2. https://github.com/doctrine/dbal/blob/de2b2ef4bb0642ebe0cb8bbc3d70fb127da338f2/src/Schema/OracleSchemaManager.php#L146-L147

But there's no hack for this for DB2, hence the failure. Most likely, we can reproduce a scenario where an integer column is mistakenly introspected as a boolean on these platforms and consider it a bug.

We could add the same hack for DB2 to satisfy the test but I'd argue that this condition in the comparator is the problem: https://github.com/doctrine/dbal/blob/de2b2ef4bb0642ebe0cb8bbc3d70fb127da338f2/src/Platforms/AbstractPlatform.php#L4021-L4021

In the context of schema migration, if two column declarations yield the same SQL, it doesn't matter if they are of the same type or not.

We could try removing this block and see if it yields any unexpected effects. Not having this condition may help get rid of the above hacks in the MySQL and Oracle platforms as well.

P.S.: doing the above would fix one of the two failures. The remaining one (where the platform unaware comparison is used) is a known limitation of the old API.

P.P.S.: given that the old comparison logic relies on the comments and it's still supported in 3.x, we can only drop the comments in 4.0 where the old API is not supported.

**Comment by greg0ire** [comment]
I'll extract the comparison loosening in a separate PR then.

File: src/Platforms/AbstractPlatform.php [hunk 1]
```diff
@@ -375,26 +375,12 @@ public function hasDoctrineTypeMappingFor(string $dbType): bool
...
-     */
-    public function getDoctrineTypeComment(Type $doctrineType): string
-    {
-        return '(DC2Type:' . $doctrineType->getName() . ')';
-    }
-
     /**
      * Gets the comment of a passed column modified by potential doctrine type comment hints.
      */
     protected function getColumnComment(Column $column): string
```
Comment (morozov) [inline src/Platforms/AbstractPlatform.php]: This method should be no longer necessary.

File: tests/Platforms/AbstractMySQLPlatformTestCase.php [hunk 2]
```diff
@@ -223,10 +223,7 @@ public function getAlterTableColumnCommentsSQL(): array
      */
     public function getCreateTableColumnTypeCommentsSQL(): array
```
Comment (morozov) [inline tests/Platforms/AbstractMySQLPlatformTestCase.php]: These methods are no longer used.

File: UPGRADE.md [hunk 3]
```diff
@@ -8,6 +8,11 @@ awareness about deprecated code.
 
 # Upgrade to 4.0
 
+## BC BREAK: Types can no longer be reverse-engineered reliably
```
Comment (morozov) [inline UPGRADE.md]: Reverse engineering is a really odd term in this context. It implies that there's no source of the data being obtained. In most cases, there is a source. I wouldn't use "reliably" either. Reliability is about something working or not working depending on a multitude of factors, usually probabilistic, while here it depends only on the column type.

I'd put it as something like "Deployed database schema no longer contains the information about abstract data types".

**Review by morozov — COMMENTED** [review]
Looks good overall, just needs some more cleanup. I believe the commits should be squashed. The first one will likely break the build.

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Looks good!

I thought a bit about the upgrade impact, and the consequences look quite straightforward. If I have a commented type in 3.x which appends its type comment to the deployed column, when I upgrade to 4.0.0, there will be a schema diff since the DBAL will no longer expect the type comment and will remove it.

Maybe the above combined with some actual schema changes will produce unexpected results but I can't think of anything right now.

**Comment by mvorisek** [comment]
With this change, what method should be overrided to add a comment about the DBAL Type to DB column?

**Comment by greg0ire** [comment]
@mvorisek I don't know that there is one.

**Comment by bohanwood** [comment]
@morozov Just a question:

I have defined a custom type, which uses DC2Type comment as of now (3.5) .

After this change, will a new migration being generated (after diffing the schema) ?
even if no schema is changed
due to Doctrine cannot correctly detect a column's type
since it's custom defined and there's no DC2Type comment.

If yes, do you have any idea to solve this in the application?

Thanks!

PS: Read #4749
> This way, the diff will be only generated if the column SQL representations on a given platform are different.

The idea seems legit. Just a little worried about how does it really work
